<html xmlns="http://www.w3.org/1999/html">
<head>
  <meta name="verify-v1" content="6wo51xu6ARUrklKNbMeXjgoTCvR8I3C7/IESIFp0t10="/>
  <link rel="stylesheet" type="text/css" href="style.css">
  <link rel="stylesheet" type="text/css" href="style.css">
  <style type="text/css">
    ol, ul {
      margin: 0.5em;
    }

    li {
      margin: 0;
    }
  </style>
  <title>Frequency-based Insilico Genome Generator</title>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
</head>
<body>
<div id="wrap">
  <div id="header"><h1>Frequency-based Insilico Genome Generator</h1></div>
  <div id="menu">
    <ul>
      <li><a href="index.shtml">Home</a></li>
    </ul>
  </div>

  <div id="sidebar">
    <h3>FIGG:</h3>
    <ul>
      <li><a href="http://sourceforge.net/projects/insilicogenome/">SF project page</a>
      <li><a href="http://sourceforge.net/projects/insilicogenome/files/">SF download page</a>
      <li><a href="https://github.com/skillcoyne/IGCSA/tree/v1.1">Repository</a>
    </ul>
    <br>

    <h3>Links:</h3>
    <ul>
      <li><a href="http://wwwen.uni.lu/lcsb/research/computational_biology">Computational Biology Group @LCSB</a></li>
      <li><a href="http://wwwen.uni.lu/lcsb">Luxembourg Centre for Systems Biomedicine</a></li>
    </ul>
    <h3>Requirements:</h3>
    <ul>
      <li><a href="http://www.oracle.com/technetwork/java/javase/downloads/index.html">Java 1.6+</a></li>
      <li><a href="http://dev.mysql.com/downloads/">MySQL 5.1</a></li>
      <li><a href="http://hgdownload.cse.ucsc.edu/downloads.html">FASTA files</a></li>
    </ul>
  </div>

  <div id="main">
    <!-- BEGIN OF THE BODY -->
    <h2>Introduction</h2>

    <p>
      FIGG is a genome simulation tool that uses known or theorized variation probability, per a given fragment size and grouped by GC content across a genome to generate
      new genomes in FASTA format while tracking applied mutations for use in analysis tools or population simulations.
    </p>
    <br>

    <h2>Installation</h2>

    <p>
      <a href="http://www.oracle.com/technetwork/java/javase/downloads/jre6-downloads-1637595.html">Java 1.6</a> or higher is required to run FIGG.
      Download and expand one of the provided compressed packages from the <a href="http://sourceforge.net/projects/insilicogenome/files/">SF download page</a>. This package
      includes a database and one FASTA file for an example run.

    <h3>Run Provided Example</h3>
    After uncompressing the package in a Unix or OSX system open a terminal and run:
    <pre>sh figg.sh -n MyGenome -o -c 22 </pre>
    on Windows run cmd.exe and in the window run:
    <pre>figg.bat -n MyGenome -o -c 22 </pre>
    Chromosome 22 takes about 26 minutes to fragment and mutate on current hardware. A
    <h3>In order to run a full genome</h3>

    <p>
    <ul>
      <li>Download human FASTA files from <a href="ftp://ftp.ncbi.nlm.nih.gov/genbank/genomes/Eukaryotes/vertebrates_mammals/Homo_sapiens/GRCh37/Primary_Assembly/assembled_chromosomes/FASTA">NCBI</a> or <a
          href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/">UCSC</a>. Note that only chromosomes 1-22, X and Y are represented in the database.
      </li>
      <li>Update the dir.assembly property in the genome.properties file to point to the directory where the FASTA files were saved</li>
      <li>Alter the dir.insilico directory if you wish to save the mutated genomes in a new location.</li>
      <li>Run the correct script for your OS (bat/sh file) without the -c option: <pre>sh figg.sh -n MyGenome -o -t 12</pre></li>
    </ul>
    <b><i>Note:</i></b> The full genome run using the thread option -t 12 takes 32-36 hours to complete on current hardware.
    </p>

    <h2>FAQ</h2>
    <dl>
      <!--<dt>How can I cite FIGG?</dt>-->
      <!--<dd>-->
      <!--The small-scale variation fragment mutation has been submitted to <a href="http://bioinformatics.oxfordjournals.org/">Bioinformatics</a>.-->
      <!--<ul>-->
      <!--</ul>-->
      <!--</dd>-->
      <dt>Does FIGG work with all genomes?</dt>
      <dd>
        Currently this has only been tested on human GRCh37. However, the tool itself should work with any genomic data in FASTA format.
        The database tables will need to be generated based on your specific genome and known/hypothesized variation information and the
        scripts available in the source download will require alteration to work for your files or new ones written.
      </dd>
      <dt>What data is provided?</dt>
      <dd>
        The data provided in the database tables were based on analysis of <a href="http://www.1000genomes.org/">1000Genomes</a> and <a href="http://hapmap.ncbi.nlm.nih.gov/">HapMap</a> variations called against human reference <a
          href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/">GRCh37 (hg19)</a>. These variation
        tables can be used with any human genome sequence, however variation frequency may change between releases.
      </dd>
      <dt>Can I include specific variations at known locations?</dt>
      <dd>
        While the code is set up to support this, the current implementation does not provide for location based, or specific genomic alterations.
        This may be added in the future.
      </dd>
    </dl>
    <h2>Usage</h2>
    <dl>
      <dd>
      <dt>Usage: figg.sh (or figg.bat) [options]</dt>
        <pre>
    -n, --name Genome directory name, if not provided a random name is generated.
    -o, --overwrite Overwrite genome directory if name already exists.
    -c, --chromosome List of chromosomes to use/mutate, comma-separated (e.g. 21,3,X). If not included chromosomes will be determined by
    the fasta files found in the dir.assembly directory (see genome.properties).
    -t, --threads Number of concurrent threads, default is 5. Each thread handles a single chromosome.
    -h, --help Print usage help.
        </pre>
      </dd>
      <dt>Examples using the provided shell script:</dt>
      <dd>
        Generate mutations for chromosome 22, overwrite previous named run.
        <pre>figg.sh -n MyChr17 -o -c 22</pre>

        Generate mutations for all chromosomes in the dir.assembly directory using 12 threads. Don't overwrite.
        <pre>figg.sh -n MyGenome -t 12</pre>

      </dd>
    </dl>
    <h2>Notes & Warnings</h2>
    <dt>Database Use</dt>
    <dd>
      The Derby database is provided within the zip package. The provided setup connects directly (see genome.properties). If you intend to use this on a cluster it is currently recommended that you set up your cluster run to copy the derby
      database directory (db) to the local node for IO speed and update the properties file appropriately.
      If you prefer MySQL please <a href="http://sourceforge.net/projects/insilicogenome/files/Databases/insilico-genome-mysql.zip/download">download the database dump</a> and alter the properties file appropriately. Note that no guarantees
      are made for speed if you choose to connect to a remote MySQL database from a cluster environment. While a local MySQL database is as fast as Derby, tests of the remote database show slowdowns of more than 50%.
    </dd>
    <h2>Software License</h2>
    <dt><a href="http://www.apache.org/licenses/LICENSE-2.0.html">Apache License 2.0</a></dt>
    <pre>
      Copyright 2013 University of Luxembourg and the Luxembourg Centre for Systems Biomedicine

      Licensed under the Apache License, Version 2.0 (the "License");
      you may not use this file except in compliance with the License.
      You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

      Unless required by applicable law or agreed to in writing, software
      distributed under the License is distributed on an "AS IS" BASIS,
      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
      See the License for the specific language governing permissions and
      limitations under the License.
      </pre>
    </p>

    <!-- END OF THE BODY-->
  </div>
  <!-- BEGIN OF THE FOOTER -->
  <div style="clear: both;"></div>
  <div id="footer">
    <hr/>
    <p style="text-align: center;">
      This work is supported by an AFR grant from the <a href="http://www.fnr.lu/">Fonds National de la Recherche Luxembourg</a>
      <br>
      Copyright 2013 <a href="http://uni.lu">University of Luxembourg</a> and the <a href="http://wwwen.uni.lu/lcsb">Luxembourg Centre for Systems Biomedicine</a>
      <br>
      Last modified: 2013-05-13</p>
  </div>
  <!-- END OF THE FOOTER -->
</div>
</body>
</html>
